core/src/main/scala/cats/syntax/applicativeError.scala

```diff
@@ -3,12 +3,18 @@ package syntax
 
 import cats.data.{Xor, XorT}
 
-trait ApplicativeErrorSyntax {
+trait ApplicativeErrorSyntax1 {
+  implicit def applicativeErrorSyntax[F[_], E, A](fa: F[A])(implicit F: ApplicativeError[F, E]): ApplicativeErrorOps[F, E, A] =
+    new ApplicativeErrorOps[F, E, A](fa)
+}
+
+trait ApplicativeErrorSyntax extends ApplicativeErrorSyntax1 {
   implicit def applicativeErrorIdSyntax[E](e: E): ApplicativeErrorIdOps[E] =
     new ApplicativeErrorIdOps(e)
 
-  implicit def applicativeErrorSyntax[F[_, _], E, A](fa: F[E, A])(implicit F: ApplicativeError[F[E, ?], E]): ApplicativeErrorOps[F[E, ?], E, A] =
+  implicit def applicativeErrorSyntax2[F[_, _], E, A](fa: F[E, A])(implicit F: ApplicativeError[F[E, ?], E]): ApplicativeErrorOps[F[E, ?], E, A] =
```

Comment:
`Unapply` is a more general solution to the type inference problem that you are currently solving with this PR. But it will (hopefull soon!) become useless because [SI-2712](https://issues.scala-lang.org/browse/SI-2712) is going to [be fixed](https://github.com/scala/scala/pull/5102).

Once this will be the case, projects like cats will follow some migration path to get rid of the current workarounds. One possible migration path could be to move every `Unapply`-related method to an external project that could be kept maintained just for compatibility with old Scala versions.

So, using `Unapply` clearly marks that the implicit definition will be useless once SI-2712 is fixed.